### PR TITLE
Issue 707 - Mongo Driver Upgrade Part 3

### PR DESCRIPTION
**Pull request by sebjf**
This PR addresses #707, performing the actual mongo driver upgrade.

TLDR;
* The mongo driver has been upgraded as well as a few improvements made to the codebase in general.
* The mongo driver on `tests` has been updated in-place, so there are no build changes for the automated pipelines.
* However, this branch relies on a specific branch of `tests`, as well as specific branches of asset_generator and svgexporter, so once this PR is approved those should all first be merged and the pointers updated here before merging, like for other PR.
* Binaries for building on Windows are available here: https://asitesol.sharepoint.com/:u:/s/3DRepoTeamFolder/EQejn9RvoAZFv7GMgVdpdakB9eXqL_GSmwXYgJGQrbme5g?e=gbGsft

## Build Changes

1. The mongocxx driver expected is now 3.11+. Both the cxx driver and bsoncxx library are required, as well as the C driver, which is a dependency of the cxx one. A new CMake module, FindMongoCXXDriver, has been added to configure the new driver, and replaces FindMongo.

2. The Boost version targetted is nominally 1.86. Boost 1.71 (which works fine) is installed on Linux via the package manager by the existing build scripts.  Prebuilt libraries are available for Windows online. A somewhat recent version of Boost is required for the new c++ standard used.

3. C++20 is now the standard specified by CMakeLists.

The FindMongoCXXDriver uses the existing MONGO_ROOT environment variable, and assumes the C and CXX driver are installed to the same location. The mongo-cxx-driver.tgz file has been updated in-place. Therefore despite the above changes, the upgrade should be transparent and not require any changes to the existing build pipelines outside what is in this repo itself.

## Major Changes

MongoDatabaseHandler, RepoBSON, RepoBSONBuilder and RepoBSON's unit tests have had their internals updated to support the new mongo API. MongoDatabaseHandler has also had its unit test coverage extended.

Additionally a couple of larger refactors were made, as they were affected by the API upgrade indirectly and it was better to make the improvements now than to put effort into backwards compatibility for things that are not ideal anyway:

### RepoController/RepoManipulator

RepoController/RepoManipulator have been refactored, as the database handling had become confused over time. For example, RepoManipulator had methods to connect to a database, but the global instance was disconnected by _RepoControllerImpl's instance destructor. This had previously caused problems in the unit tests, so a refactor was overdue. 

Now, each RepoManipulator holds its own instance of MongoDatabaseHandler. Unused parameters have been removed from the RepoManipulator API. Representing database credentials as BSON is no longer supported; where user parameters such as the database username need to be extracted, they are provided by the RepoToken directly.

RepoController no longer disconnects explicitly. The database handler is assumed to clean up after itself when it is destructed. The lifetime of the handler is managed by smart pointers. This approach is consistent with the intended use of Mongo's pools and clients. 

The custom mongo connection pool has been completely removed - pools are now provided as part of the driver.

### AbstractDatabaseHandler/FileManager

Following the above, the FileManager management has been updated too. The relationship between MongoDatabaseHandler and FileManager is a little tricky, since both rely on each other, but both have public APIs (i.e. the FileManager is passed itself to SceneManager methods, and is used directly to store files, such as the RepoBundles). It is not clear which should own the other, and holding shared pointers to each other will result in a circular reference.

In the end, MongoDatabaseHandler has been given a shared pointer to FileManager, and FileManager holds a weak pointer to MongoDatabaseHandler. This is because MongoDatabaseHandler can function, in a reduced way, without FileManager, but not vice-versa. A FileManager is initialised and assigned to the MongoDatabaseHandler instance by RepoManipulator when RepoManipulator is initialised.

Since the use of files is an implementation detail, the FileManager reference is held by MongoDatabaseHandler, not AbstractDatabaseHandler. 

Managers are stored as shared pointers, though are passed to SceneManager etc as regular pointers, as updating the SceneManager APIs is out of scope, and for now all usage is synchronous.

BlobFilesHandler now takes a shared pointer to the FileManager, as this is best practice, if not strictly necessary given how it's used at the moment.

### Expressions

A new set of classes has been introduced to encapsulate expressions (that is, queries, indices and filters). These sit in the repo::core::handler::database namespace, as they are for use with the database API.

Internally, they use RepoBSONBuilder to build RepoBSONs, which in turn create bsoncxx documents, so they support all the additional Repo type overloads (UUIDs, etc).

The primary requirement was to replace Mongo macros that no longer exist. However the new types also fix abstraction leaks (specifically, RepoScene built a complex query that required it to know Mongo's syntax), add additional type safety (the conversion to RepoBSON is private via friend classes), and can be unit tested (unit tests are part of the MongoDatabaseHanlder suite). Finally while Mongo API uses bsoncxx documents now, they also allow replacing the actual Mongo expression object with something else if in the future Mongo makes a more performant option available. For example, mongo already now supports a dedicated type for index options, should we wish to use them.

Currently, enough functionality has been added to support RepoScene's loadRevision query and index construction. (In practice almost all other queries can use the dedicated findOneByUniqueID or findOneBySharedID convenience methods).

### Error Handling

The new mongo driver leans into the distributed design pattern and has no capacity to check the integrity of a connection - it will simply throw if there is any problem.

The MongoDatabaseHandler unit tests have been updated to perform an operation explicitly against the database, to test that the parameters are being passed correctly. Whether an AbstractDatabaseHandler pointer is valid or not is no longer an indicator of whether the database is available. 

However, RepoManipulator will also explicitly attempt an operation in connectAndAuthenticateWithAdmin to verify that the parameters can facilitate a connection at least once before proceeding, so that part of the error handling in the client should work as before.

Otherwise, the MongoDatabaseHandler now no longer catches exceptions as a rule, but rather irrecoverable exceptions are expected to bubble up to the client where they can be clearly printed. This is because in practice none of our operations are really 'optional', where a failure could be recovered from, so it is better to fail fast and early and minimise any garbage created.

The exception behaviour for MongoDatabaseHandler for invalid databases is as follows: Write operations (such as insertDocument) will throw, and read operations will return the default value. This is because Mongo's new intended usage is for lazy initialisation (so, inserting to a non-existent database creates that database, and reading from a non-existent one returns nothing); invalid names cannot lazy initialise so throw. Reading however behaves as if the database/collection was empty. For the same reason, Drop operations will also not throw, because even though they are sort-of writes, they are not explicit (i.e. calling them on empty databases, etc does not throw).

### RepoBounds

To continue the move to dedicated types for complex schema, bounding boxes are now represented by a new type, RepoBounds. This uses RepoVector3Ds underneath. Bounds are always 64 bit in memory and the database (which is consistent with how they were stored before).

The new type includes convenience functions to update the bounds, replacing copies of the same algorithm that were duplicated across different parts of the codebase prior.

## Other Modifications

1. The use of RepoBSONBuilder has been extended further, now being used by RepoBSON internally in place of the mongo types for building the binaries references.

2. RepoBSONBuilder has been given an explicit overload for std::vector, as the new Mongo API no longer accepts this for arrays.

3. Header files have been improved similarly to Part 2, where ever they were otherwise touched.

4. Changed Boost `leaf` method to `filename` (https://www.boost.org/doc/libs/1_45_0/libs/filesystem/v3/doc/deprecated.html)

5. A new findOneByUniqueID overload was added to AbstractDatabaseHandler to take ids-as-strings. Anywhere findOneByCriteria was called with an object that was actually just matching the _id field as been updated to use findOneByUniqueID.

6. A new connection options struct has been introduced for MongoDatabaseHandler, supporting timeouts as well.

7. Admin database definitions and methods have been removed because they are not used anywhere.

8. findAllByUniqueIDs has been removed as it is not used anywhere (this was never actually part of AbstractDatabaseHandler)

9. It is no longer possible to de-serialise a SupermeshNode. This was only supported for legacy purposes before, but it is getting troublesome to keep supporting it, and realistically its hard to imagine any scenario where we'd want to read supermesh nodes in bouncer, rather than just re-process the scene.

10. _RepoVector3D now has an explicit conversion (constructor) to go from doubles to floats.

11. RepoBSON::getFloatVectorField has been removed as it is unused and the BSON spec doesn't store floats natively (getDoubleVectorField is a drop-in replacement that should be used instead, and has been in the unit tests).

12. RepoBSONElement::binData has been removed. This is incompatible with the new drivers memory model, and was never used publicly (i.e. outside of RepoBSON/RepoBSONBuilder).

13. Other RepoBSONElements (eoo, isNull and the implicit and explicit string operations) have also been removed, as they were not used, and in the future should not be used (instead such parsing should happen inside RepoBSON direct against the bsoncxx types).

14. The bsoncxx library no longer throws when building large documents, since the 16Mb limit is a mongo specific limit, not a BSON one. The unit test has been removed from RepoBSON. Size checks should performed by the nodes themselves that may expect to go over the limit (like the sequences).

### RepoBSONBuilder

RepoBSONBuilder inherits from the new core builder of bsoncxx. There are no major API changes. Internally, the changes are as follows:

1. Key and values are appended in different operations.
2. Arrays are built through the open_array & close_array operations (instead of making new objects and appending them as before).

This is much simpler than before and should also provide improved performance since the entire BSON is built 'in-place' and reduces the amount of memory moved around.

3. The methods that convert non-native types (UUID, tm, etc) are now overloads rather than specialisations. This is because the rules around overloads are much less ambiguous than for specialisations, and the only benefit to specialisations in this case are that they don't need to be declared in the header, and ours were previously anyway.

### RepoBSON

RepoBSON continues to use RepoBSONElement for simple types, but makes more use of the privately inherited document methods for complex schema such as matrices, and arrays. This is because the way that the memory ownership in the new API works makes this a neater and more peformant implementation.

### MongoDatabaseHandler

MongoDatabaseHandler now uses the drivers implementation of pools, and a pool is held per instance, making it easier to implement multithreading in the future.

For queries and indices, dedicated types are used for type-safety, even though internally these still turn into BSON objects for now.

More of the query schema has been moved inside MongoDatabaseHandler.

The Mongo APIs to get databases and collections and perform operations have changed (obviously).

There is some explicit logic to emulate our desired behaviour if e.g. invalid database or collection names are provided, for which the driver behaviour now differs.

## Repo API Changes

Ideally, there would be no public API changes for any of our types (RepoBSON, RepoBSONBuilder), but for performance reasons or other there are a few concessions made:

1. unsigned integers, size_t or similar, are now cast to int32_t whenever they are appended to a BSON (i.e. in the RepoNode, etc). This is because the BSON spec does not actually support these unsigned types. Previously they were internally cast to int32_t by mongo. While we could make RepoBSONBuilder backwards compatible, as these casts are potentially lossy it is much better practice to make them explicit where they are performed. Similarly, all instances of `long long` have been replaced with `int64_t` as `long long` doesn't actually specify a specific bit width on gcc, so `int64_t` is more descriptive.

2. The nFields method has been removed. This is because there isn't an obvious replacement, and we'd have to iterate over the document to get the count. Since almost every user of this method then iterates over the fields anyway, the method has been deprecated to encourage the caller to use the most performant implementation when parsing BSONs.

3. Getting an Array of fields (RepoBSONElements) is no longer supported. Now the assumption is that callers know the type ahead of time. This is because the new driver makes much more use of views, which do not own the underlying memory, so it is not safe to pass around RepoBSONElements as before. If we wanted something like this in the future, we should use RepoVariants, which maintain their own memory - this is more costly though, so the current refactor has just avoided the need to pass element references at all.

4. It is no longer supported to connect to a local database without a username and password. Accordingly, the default arguments for username and password have been removed from the constructor and getHandler.

5. It is no longer possible to build, or get, an "array document" (a BSON with fields that are indices) in place of an array. This is because the mongo API no longer accepts this schema. When reading arrays, they must be returned as strongly typed vectors. Similarly, it is no longer possible to get arrays of objects (documents) as "array documents" and they must be read using the new getObjectArray method. This is perhaps the most significant change overall. While it would be possible to emulate the old API, the new one allows both reading and writing arrays, including nested ones, using true array container types, which should be preferred as it is more concise, performant and offers more compile time checks.

6. getIntField will no longer silently convert doubles or other numbers that cannot be losslessly converted. All callers of getIntField have been checked to make sure they are not assuming this will happen.

Finally,

Not so much a change since we never explicitly tested for this, but the equality operator only returns true if the fields are in the same order for each BSON, which is not perhaps intuitive and something to note.

**Tests** 

A number of manual tests similar to those on 703 were run.

There was one bug found:

1. On Windows the unit literals for IFCs were being written as an encoding that was not utf8. A CMake fix was made, and a unit test added to capture this in the future.

### CODE REVIEW

**Comment by sebjf** [comment]
The instructions to build the driver on Windows are here: 

https://www.mongodb.com/developer/products/mongodb/getting-started-mongodb-cpp/
https://www.mongodb.com/docs/languages/cpp/cpp-driver/current/installation/windows/#std-label-cpp-installation-windows

The versions used are mongoc 1.19 and mongocxx 3.11

They are repeated below for completeness:

1. Download the mongoc driver and extract to where the bouncer libraries should live.
2. Create two directorues, `_build` & `install`, and enter `_build`.
2. Configure the build like so: 
	
	`cmake -G "Visual Studio 16 2019" -A x64 -S .. -B . -DCMAKE_INSTALL_PREFIX=D:\3drepo\bouncer\mongo-c-driver-1.29.0\install`
	
3. Set the Configuration to RelWithDebugInfo and Build All and INSTALL

4. Download the mongocxx driver and extract to where the bouncer libraries should live
5. Create two directorues, `_build` & `install`, and enter `_build`.
6. Configure the build like so:
	
	`cmake -G "Visual Studio 16 2019" -A x64 -S .. -B . -DCMAKE_INSTALL_PREFIX=D:\3drepo\bouncer\mongo-cxx-driver-r3.11.0\install -DCMAKE_CXX_STANDARD=20 -DCMAKE_PREFIX_PATH=D:\3drepo\bouncer\mongo-c-driver-1.29.0\install -DBOOST_ROOT=D:\3drepo\bouncer\boost_1_86_0`
	
7. Give the following commands:

	`cmake --build . --config RelWithDebInfo`
	`cmake --build . --target install --config RelWithDebInfo`


Note the following about the above:

1. The Boost version must be the same as used to build bouncer (or, two sets of Boost libraries will need to be accessible on the Path)
2. The same version of Visual Studio used to build bouncer should be specified as the generator.


When building and running bouncer:

1. The MONGO_CXX_DRIVER_ROOT CMake variable must point to the CXX driver install directory. CMake for bouncer does not need the C driver.
2. When runnning bouncer, both the C and CXX shared libraries must be available on the Path.

Finally, 3drepoSynchroReader 4.0.2 still relies on Boost 1.59 vc140, so unless this has been rebuilt, those older Boost libraries must still be available.

A set of binaries for both mongo drivers are available here: https://asitesol.sharepoint.com/:u:/s/3DRepoTeamFolder/EQejn9RvoAZFv7GMgVdpdakB9eXqL_GSmwXYgJGQrbme5g?e=gbGsft

**Review by carmenfan — CHANGES_REQUESTED** [review]

File: bouncer/src/repo/core/handler/repo_database_handler_mongo.cpp [hunk 1]
```diff
@@ -140,57 +114,40 @@ bool MongoDatabaseHandler::caseInsensitiveStringCompare(
...
+	auto client = clientPool->acquire();
+	auto db = client->database(database);
+	if (!db.has_collection(name)) {
 
-	return authBson;
-}
+		// This is an explicit operation that will throw if a collection already
+		// exists. Note that for writing documents, collections are created
+		// explicitly so this is only needed to create empty collections, such as
+		// the issues list.
```
[the remark this answers is not in the document]
Comment (sebjf) [reply bouncer/src/repo/core/handler/repo_database_handler_mongo.cpp:126]: 😆 in that case i'll remove it, since the driver is explicitly designed to do lazy initialisation of collections!

File: bouncer/src/repo/manipulator/repo_manipulator.cpp [hunk 2]
```diff
@@ -50,54 +55,55 @@ RepoManipulator::~RepoManipulator()
...
+		username,
+		password,
+		options
+	);
+	try {
+		dbHandler->getCollections(ADMIN); // Test the connection to the database
+	}
+	catch (std::runtime_error)
+	{
+		return false;
```
[the remark this answers is not in the document]
Comment (sebjf) [reply bouncer/src/repo/manipulator/repo_manipulator.cpp:79]: Agreed, I have thought some more and tweaked how this works.

`connectAndAuthenticateWithAdmin` no longer returns anything but calls the new testConnection method. The exception (if any) bubbles up to the client, where it is caught and the error code (because the mongo exceptions inherit from system_error, rather than exception) are used to infer coarsely the type of error for the return code.

The original exception object is used to print the error so those details are in the log.

While the use of error codes is an implementation detail, the alternatives are to include the mongo exception headers in the client, which is leaky, or to catch and wrap the exceptions in RepoException.

We have done the latter in places, but I am not too fond of this idea as (a) it is potentially lossy (the places we do it explicitly add context such as bson field names, but in this case it would be more likely to lose it) and (b) more importantly, if we assume we always wrap native exceptions, and miss one, it won't be handled correctly.

My feelings are that these should be very rare and fatal conditions, so its worth just letting the native ones rise and enter the log unadultered, and if we can get the right return code great, but it should be secondary to maintainability.

File: bouncer/src/repo/manipulator/repo_manipulator.cpp [hunk 3]
```diff
@@ -50,54 +55,55 @@ RepoManipulator::~RepoManipulator()
...
```
[the remark this answers is not in the document]
Comment (sebjf) [reply bouncer/src/repo/manipulator/repo_manipulator.cpp:104]: ditto above!

File: bouncer/src/repo/core/handler/fileservice/repo_file_manager.cpp [hunk 4]
```diff
@@ -349,18 +329,10 @@ bool FileManager::upsertFileRef(
...
 	auto refObj = makeRefNode(id, link, type, size, metadata);
 	std::string collectionName = collectionNamePrefix + "." + REPO_COLLECTION_EXT_REF;
-	success = dbHandler->upsertDocument(databaseName, collectionName, refObj, true, errMsg);
-	if (!success)
-	{
-		repoError << "Failed to add " << collectionName << " file ref: " << errMsg;;
-	}
-
-	return success;
+	getDbHandler()->upsertDocument(databaseName, collectionName, refObj, true);
```
[the remark this answers is not in the document]
Comment (sebjf) [reply bouncer/src/repo/core/handler/fileservice/repo_file_manager.cpp:334]: This will throw on any database errors.

If there is an exception inserting the ref node, that is a fatal error, so the exception is caught in the client and the whole process terminated.

We could catch the exception and delete the file we just wrote, but in that case there is still the question of what to do next: it may be half the scene has already been written, and if there is a database failure the import can't continue. It seemed best since we can't guarantee to clean up properly, to just let the process fail and the maintenance scripts handle it when they next run.

[the remark this answers is not in the document]
Comment (carmenfan) [reply bouncer/src/repo/core/handler/fileservice/repo_file_manager.cpp:334]: I'm less worried about orphan data, but I'm more thinking this will make the logs more cryptic and hard for us to know what happened based on it when we need to diagnose an issue

Unless they've significantly improved their message details, the mongo error will only typically tell you the exact error (e.g. duplicated id) and we wouldn't really know what was the bouncer action that trigged it, or which collection we were working with etc.

We may be able to guess with the other logs surrounding it on the stage, but there are situations where we're doing a bulk of operations I think and it may be helpful to get more insights on error 🤔 

so even if it's just a catch just to log some details about what it was trying to do then throw it again it may be useful? (whether it's in here or at the caller)

**Comment by carmenfan** [comment]
> Finally, 3drepoSynchroReader 4.0.2 still relies on Boost 1.59 vc140, so unless this has been rebuilt, those older Boost libraries must still be available.

@sebjf  I don't know if you want to do this here or make it a separate issue to upgrade, but we do own the Synchro compilation process and I think it was me who added boost in (for timezone and file reading purposes .... I think)

So we could probably remove the 1.59 dependency there too.

File: bouncer/src/repo/core/handler/repo_database_handler_mongo.cpp [hunk 5]
```diff
@@ -701,157 +424,109 @@ bool MongoDatabaseHandler::insertDocument(
...
-		{
-			std::string errString(e.what());
-			errMsg += errString;
-		}
-	}
-	else
-	{
-		errMsg = "Unable to insert Document, database(value : " + database + ")/collection(value : " + collection + ") name was not specified";
+		repoInfo << "Inserting " << toCommit.size() << " documents...";
+		col.insert_many(toCommit);
```
[the remark this answers is not in the document]
Comment (sebjf) [reply bouncer/src/repo/core/handler/repo_database_handler_mongo.cpp:460]: Yes the driver still throws (in fact it seems it's its only way of providing feedback!)

When I was looking at the callers, it appears all the operations are effectively "required" in the sense that a failure of any of them would leave the, e.g. scene, in an incomplete state. Therefore, the best thing to do in encountering an unexpected failure is to bail out, like before, but this way through letting the exception bubble up to the client where it can be printed to the log with all the information in the original exception object.

(The exception being where we catch the exception object in order to rethrow with additional context, such as bson field names during parsing.)

**Comment by sebjf** [comment]
> > Finally, 3drepoSynchroReader 4.0.2 still relies on Boost 1.59 vc140, so unless this has been rebuilt, those older Boost libraries must still be available.
> 
> @sebjf I don't know if you want to do this here or make it a separate issue to upgrade, but we do own the Synchro compilation process and I think it was me who added boost in (for timezone and file reading purposes .... I think)
> 
> So we could probably remove the 1.59 dependency there too.

I will have a look at it! 

I mainly mentioned it just so you and Felix wouldn't get super annoyed wondering what was still linking against the old Boost (Dependencies was particularly unhelpful, because it didn't jump to the missing load, and I had to go expand and search the tree by hand! I was convinced it was mongo for half an hour...), but actually since we are here it is worth at least checking, as maybe it might cause problems down the line if things start diverging...

File: bouncer/src/repo/manipulator/repo_manipulator.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (carmenfan) [reply bouncer/src/repo/manipulator/repo_manipulator.cpp:79]: I think it is better if we catch the mongo exceptions here, where we're expected it understand them, and translate them to RepoException or just print the message with additional context of what we're trying ot do then carry on with whatever we're trying to do.

Whilst this doesn't happen often when something is up and running, we do run into this a lot when we're trying to setup a deployment and helps our diagnosis. So I think I'd prefer it if we can handle it with as much interpretation as possible.

Again this could be like the the existing code, we keep the boolean check return, but just log the error here in the log with some wrapper text on what we were trying to do.

File: bouncer/src/repo/core/handler/fileservice/repo_file_manager.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (sebjf) [reply bouncer/src/repo/core/handler/fileservice/repo_file_manager.cpp:334]: Makes sense! I will rework the RepoExceptions to hold more information, such as nested exceptions, and at each stage outwards catch and rethrow where there is important context so by the time it gets to the outer handler we have information from throughout the stack. I'll update the connection testing to do the same too.

**Comment by sebjf** [comment]
Hi @carmenfan, I've updated the error handling.

I have set it up to use [throw_with_nested](https://en.cppreference.com/w/cpp/error/throw_with_nested), allowing native exceptions to be wrapped by our own which can provide additional context, using the language features so we don't need to maintain complex objects.

For MongoDatabaseHandler, it has its own exception type (MongoDatabaseHandlerException) that adds the operation, database, collection and URI to all exceptions. RepoException has a member that can print all nested exceptions recursively, which is used by 3drepobouncerClient in the outermost handler.
Additionally, RepoException has been given a member allowing exception subtypes to define the error code to be used as the return value, so this can be set when the exception is created, instead of inferred in the handler.

So now, for example, RepoManipulator calls testConnection as before, with no explicit error handling:

```
void RepoManipulator::connectAndAuthenticateWithAdmin(
	const std::string& address,
	const uint32_t& port,
	const uint32_t& maxConnections,
	const std::string& username,
	const std::string& password
)
{
	repo::core::handler::MongoDatabaseHandler::ConnectionOptions options;
	options.maxConnections = maxConnections;
	dbHandler = repo::core::handler::MongoDatabaseHandler::getHandler(
		address,
		port,
		username,
		password,
		options
	);
	dbHandler->testConnection();
}
```

But if called with invalid credentials, the log will show (dropping duplicate lines):

```
Logging directory is set to ./log/
[04-12-2024 01:49:01][INFO]: 3D Repo Bouncer Version: 5.12_1
[04-12-2024 01:49:01][INFO]: Operation: import
[04-12-2024 01:49:01][INFO]: Arg 1: D:\3drepo\tests\cplusplus\bouncer\data\models\cube.obj
[04-12-2024 01:49:01][INFO]: Arg 2: sandbox
[04-12-2024 01:49:01][INFO]: Arg 3: cube
[04-12-2024 01:49:11][ERROR]: MongoDatabaseHandler exception: testConnection() failed.
[04-12-2024 01:49:11][ERROR]:   MongoDatabaseHandler exception: in getCollections on: admin Uri: mongodb://testUser:3drepotest@invalidAdd:12345/?maxConnecting=1&socketTimeoutMS=10000&serverSelectionTimeoutMS=10000&connectTimeoutMS=10000
[04-12-2024 01:49:11][ERROR]:           No suitable servers found: `serverSelectionTimeoutMS` expired: [Failed to resolve invalidadd]. Topology type: Single: generic server error

D:\3drepo\3drepobouncer_ISSUE707\3drepobouncer\build\client\RelWithDebInfo\3drepobouncerClient.exe (process 63808) exited with code 2.
```

And the return value is 2 (REPOERR_AUTH_FAILED), which is the default code for database exceptions.


Or, for example, when reading the RepoConfig, fromFile is wrapped in a try..catch which returns a nested RepoConfigException object. The client no longer has to perform a distinct validation step, but can just do:

```
auto config = repo::lib::RepoConfig::fromFile(configPath);
repo::RepoController::RepoToken* token = controller->init(errMsg, config);
```

And if the config has a json problem, the log looks like:

```
Logging directory is set to ./log/
[04-12-2024 02:10:21][INFO]: 3D Repo Bouncer Version: 5.12_1
[04-12-2024 02:10:21][INFO]: Operation: import
[04-12-2024 02:10:21][INFO]: Arg 1: D:\3drepo\tests\cplusplus\bouncer\data\models\cube.obj
[04-12-2024 02:10:21][INFO]: Arg 2: sandbox
[04-12-2024 02:10:21][INFO]: Arg 3: cube
[04-12-2024 02:10:30][ERROR]: Exception reading RepoConfig: D:\3drepo\tests\cplusplus\bouncer\data\models\config\badConfig1.json
[04-12-2024 02:10:30][ERROR]:   D:\3drepo\tests\cplusplus\bouncer\data\models\config\badConfig1.json(18): expected '}' or ','

D:\3drepo\3drepobouncer_ISSUE707\3drepobouncer\build\client\RelWithDebInfo\3drepobouncerClient.exe (process 64712) exited with code 28.
```

(The innermost message coming directly from Boosts parser.)

Or if it fails our schema checks, it looks like:

```
Logging directory is set to ./log/
[04-12-2024 02:11:28][INFO]: 3D Repo Bouncer Version: 5.12_1
[04-12-2024 02:11:28][INFO]: Operation: import
[04-12-2024 02:11:28][INFO]: Arg 1: D:\3drepo\tests\cplusplus\bouncer\data\models\cube.obj
[04-12-2024 02:11:28][INFO]: Arg 2: sandbox
[04-12-2024 02:11:28][INFO]: Arg 3: cube
[04-12-2024 02:11:30][ERROR]: Exception reading RepoConfig: D:\3drepo\tests\cplusplus\bouncer\data\models\config\badConfig2.json
[04-12-2024 02:11:36][ERROR]:   Cannot find entry 'db' within configuration file.

D:\3drepo\3drepobouncer_ISSUE707\3drepobouncer\build\client\RelWithDebInfo\3drepobouncerClient.exe (process 53212) exited with code 28.
```

Again, the outer exception sets the return value, which now is REPOERR_INVALID_CONFIG_FILE (28).

So, if the native code gives us a useful message (like the json parser one), we keep it, and if its useless (like the previous mongo ones) we can get context, without having to explicitly handle errors.

We can add these try..catch blocks anywhere there may be important context. For RepoScene I have left it mostly alone because the repoTrace calls already provide context and refactoring that is out of scope. Where methods return booleans (i.e. in the FileManager), exceptions are caught explicitly, printed in full, and control handed to the existing logic.

I have also tested I haven't broken the existing error handling in places we haven't touched, by e.g. breaking a node and running genStash to see what the log looks like, and it looks OK!

**Comment by sebjf** [comment]
@carmenfan,

> we do own the Synchro compilation process and I think it was me who added boost in (for timezone and file reading purposes .... I think)
> So we could probably remove the 1.59 dependency there too.

I had a quick look and it seems Boost is also a dependency of Thrift! 

Since the versioning is not presenting a problem on Linux, I am inclined to leave it for now...

The "problem" on Windows is that the library its looking for has the boost version and compiler in the filename, whereas Linux uses the ABI version for shared object names, so seems to work fine with whatever travis ends up installing...

**Review by carmenfan — DISMISSED** [review]
@sebjf looks good and works on my laptop. If you want to merge the dependencies in then this is good to merge!
